### PR TITLE
feat: upgrade git-warp to v14.16.2

### LINKED ISSUE #312: dependency: audit and upgrade @git-stunts/git-warp before major Hill 1 implementation
## Summary
Audit the current Git Mind dependency boundary against `@git-stunts/git-warp`, then upgrade from the current 11.5.0 baseline to a supported current version before substantial new Hill 1 implementation work proceeds.

## Context
Git Mind currently declares `@git-stunts/git-warp` as `^11.5.0` and the lockfile/install are at `11.5.0`, while the current published version is materially newer. We do **not** want to build a large amount of new behavior on an aging substrate without understanding the upgrade surface first.

## Scope
- inventory the Git Mind API surface used from git-warp
- identify breaking changes and notable new capabilities across major versions since 11.5.0
- define a safe upgrade target
- add or strengthen tests that protect Git Mind's actual dependency boundary
- upgrade package metadata and lockfile
- resolve any runtime, contract, or test breakage
- clean up stale contributor docs about local-path dependency usage if still present

## Acceptance Criteria
- Git Mind's git-warp dependency surface is documented
- the upgrade target version is justified
- tests covering the dependency boundary are in place and green
- Git Mind installs and tests cleanly against the upgraded version
- README / contributor docs reflect the real dependency model

## Why Now
This is an enabling cycle. The goal is explicitly to avoid layering major new Hill 1 behavior onto an outdated git-warp baseline without first revalidating the substrate.

### PR DESCRIPTION
**Pull request by flyingrobots**
# PR Title

feat: upgrade git-warp to v14.16.2

## Summary

- upgrade `@git-stunts/git-warp` from `11.5.0` to `14.16.2`
- add a property-bag compatibility layer and apply it across Git Mind's git-warp boundary
- harden observer, epoch, review, export, merge, and content handling for the v14 object-shaped property model
- align contributor/docs guidance with the published-package dependency model and Node 22 runtime
- extend regression coverage for compatibility edge cases and normalize CLI contract test timeouts

## Problem Statement

- Git Mind was three major versions behind on `@git-stunts/git-warp` and was still coded against the older Map-shaped `getNodeProps()` behavior.
- That left the repo planning new work on top of a stale substrate and made the upgrade risk hard to reason about.
- This PR moves the runtime to the current v14 line, documents the boundary, and locks in the compatibility expectations with tests before Hill 1 implementation work continues.

### Relevant ADR(s)

- [ ] ADR-0002 (Worktree Independence and Materialization Architecture)
- [ ] ADR-0003 (Graph-Native Content, Deterministic Materialization, and Workspace Bridge)
- [x] None

### Compliance Declaration

- [x] This PR is fully compliant with all checked ADRs.
- [ ] This PR intentionally deviates from one or more checked ADRs (complete Exception Request below).

### Exception Request (Required if deviating)

- **ADR clause(s) deviated from:** None
- **Why deviation is necessary now:** N/A
- **Risk introduced by deviation:** N/A
- **Mitigation and rollback plan:** N/A
- **Follow-up ADR/issue to reconcile architecture debt:** N/A

---

### Canonical Truth & Context

- [x] Graph remains canonical truth (no dual truth with generated files).
- [x] No hidden worktree coupling introduced in core/domain/materialization paths.
- [x] Context-sensitive behavior is explicit (`--at`, `--observer`, `--trust`) or deterministically defaulted.
- [x] Resolved context is surfaced in output metadata where applicable.

### Determinism & Provenance

- [x] Pure query/materialization paths remain deterministic for identical inputs.
- [x] Mutations/materializations include provenance receipts/envelopes where required.
- [x] Cache keys (if used) are derived only from semantic inputs + pinned versions.

### Artifact Hygiene

- [x] No forbidden generated artifact paths are tracked.
- [x] Any generated artifacts intentionally tracked are in allowlisted publish paths only.
- [x] Pre-commit/CI policy checks updated or confirmed valid.

### Contracts & Compatibility

- [x] Machine-facing outputs are schema-versioned.
- [x] Breaking contract changes include version bump + migration notes.
- [x] Backward compatibility impact is documented below.

### Extension/Effects Safety (if applicable)

- [x] Extension behavior does not bypass capability restrictions.
- [x] Effectful operations use explicit plan/apply semantics and emit receipts.
- [x] Timeouts/resource bounds are defined for new script/effect paths.

---

## Scope Control

- [x] PR is single-purpose/cohesive (no unrelated refactors).
- [x] Any non-essential refactor is split into separate PR(s) or explicitly justified.

## Backward Compatibility

- **CLI/API contract changes:** No intentional CLI schema change. Existing commands now tolerate both Map-shaped and object-shaped property bags at the git-warp boundary.
- **Data model/storage changes:** No Git Mind graph schema change. Dependency runtime moved from git-warp 11 to 14.
- **Migration required?:** No user migration step beyond reinstalling dependencies and using Node 22.
- **User-facing behavior changes:** Minimum runtime is now Node.js 22+. Observer resolution and malformed epoch/review records now fail or skip defensively instead of flowing through undefined values.

### Unit

- [x] Added/updated tests for changed logic
- Commands:

```bash
npm test -- test/prop-bag.test.js test/resolve-context.test.js test/review.test.js test/epoch.test.js test/export.test.js test/contracts.integration.test.js
```

### Integration

- [x] Added/updated integration tests
- [x] Commands:

```bash
npm test -- test/contracts.integration.test.js
```

### Determinism

- [x] Determinism assertions included for relevant paths
- [x] Method: existing diff/materialization/integration suites exercised against the upgraded dependency boundary
- [x] Commands:

```bash
npm test -- test/diff.test.js test/graph.test.js test/import.test.js test/merge.test.js
```

### Contract/Schema

- [x] Schema validation updated/passing
- [x] Commands:

```bash
npm test -- test/contracts.integration.test.js test/contracts.test.js
```

### Policy Gates

- [x] Mechanical architecture gates pass
- [x] Commands:

```bash
npm run lint
npm test
```

---

## Security / Trust Impact

- [x] Threat surface changed?: Minimal; dependency upgrade only. Defensive validation now rejects malformed observer/epoch/review inputs earlier.
- [x] Trust policy impact: None.
- [x] Provenance/audit impact: Review and epoch history now skip malformed records instead of emitting partially undefined data.
- [x] New failure modes introduced: Observer configs missing `match` now fail fast with an explicit error.

### Performance Impact

- [x] Hot path affected?: Slightly; property access now goes through compatibility helpers.
- [x] Expected impact (latency/memory/io): Negligible. One Buffer copy was removed for Buffer-based content reads.
- [x] Benchmarks or profiling evidence: Full suite and CLI contract canaries remained green after the upgrade; CLI timeout budget was normalized to avoid false negatives.

### Observability / Debuggability

- [x] Errors are actionable and include context.
- [x] Logs/diagnostics added or updated where needed.
- [x] git mind status / diagnostics updated if writeback/eventing behavior changed.

## Operational Notes

- [x] Feature flag (if any): None.
- [x] Rollback strategy: Revert this PR and reinstall dependencies to return to git-warp 11.5.0.
- [x] Operational caveats: Contributors now need Node.js 22+ for local development on this branch line.

## Linked Issues / Milestones

- [x] Closes #312
- [x] Milestone: None (IBM Hills/Playbacks govern planning in this repo)

---

### MUST (Hard Gates)

- PASS

### SHOULD (Quality)

- PASS

### Verdict

- APPROVE

### CODE REVIEW

**Comment by flyingrobots** [comment]
@coderabbitai Thanks. Two dependency-model findings on this PR are stale after the governance/docs alignment that already landed on : the published-package model is intentional, and the current branch already reflects it consistently in , , , , and . The actionable feedback on the runtime boundary has been fixed in the latest commits; the local-path recommendation should be treated as a mix-up, not as the target state for this repo.

**Comment by flyingrobots** [comment]
@coderabbitai review please 🥕

File: package.json [hunk 1]
```diff
@@ -26,7 +26,7 @@
     "format": "prettier --write 'src/**/*.js' 'bin/**/*.js'"
   },
   "dependencies": {
-    "@git-stunts/git-warp": "^11.5.0",
+    "@git-stunts/git-warp": "^14.16.2",
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply package.json:29]: Thanks. This is the same stale dependency-model false positive.

The governed setup for this repository is the published npm package, not a local file: dependency:
- package.json:29 declares @git-stunts/git-warp ^14.16.2.
- package.json:30 declares @git-stunts/plumbing ^2.8.0 directly.
- GUIDE.md:67 and CONTRIBUTING.md:58-60 both describe the published-package install path.

That model is intentional and was already locked in during the docs/governance cleanup, so no package.json change is required here.

File: GUIDE.md [hunk 2]
```diff
@@ -64,7 +64,7 @@ cd git-mind
 npm install
 ```
 
-git-mind depends on `@git-stunts/git-warp`, which is currently installed from a local path. See [CONTRIBUTING.md](CONTRIBUTING.md) for development setup details.
+git-mind depends on the published `@git-stunts/git-warp` package. See [CONTRIBUTING.md](CONTRIBUTING.md) for current development setup details.
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply GUIDE.md:67]: Thanks. This is a stale dependency-model mix-up, not a live contradiction.

Current canon is the published-package model:
- GUIDE.md:67 states Git Mind depends on the published @git-stunts/git-warp package.
- CONTRIBUTING.md:58-60 requires Node.js >= 22 and npm access to install published dependencies, including @git-stunts/git-warp.
- package.json:29-30 declares @git-stunts/git-warp ^14.16.2 and @git-stunts/plumbing ^2.8.0 directly.

AGENTS.md and CLAUDE.md were already updated earlier in this cycle and no longer contain the old local-path requirement. So there is no docs mismatch left on this branch; this thread is based on stale prior guidance.

File: src/prop-bag.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ * @param {PropBag} props
+ * @returns {number}
+ */
+export function getPropSize(props) {
+  if (!props) return 0;
+  if (typeof props.size === 'number') {
+    return props.size;
+  }
+  return Object.keys(props).length;
+}
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply src/prop-bag.js:54]: Thanks. This one is already fixed on the current branch.

Current code in src/prop-bag.js only treats real Map instances as size-authoritative:
- getPropSize(props) returns props.size only when props instanceof Map.
- Plain objects fall back to Object.keys(props).length.

The regression is also covered in test/prop-bag.test.js by the case that uses a plain object with { get, entries, size, status } and asserts getPropSize(props) === 4. So this thread is stale against the current diff.
